### PR TITLE
Add Bal Library Url expansion variable

### PR DESCRIPTION
**Pull request by elliotcmorris**
#86 

Add Bal Library Url expansion variable.
Perform path normalization on any path that begins with "File:\"

The intent of this is so folks can take BAL locations and use them in the path/url converters openassetio provides.

This path stuff boggles me a bit, i'm not sure if the testing is sufficient.

### CODE REVIEW

File: plugin/openassetio_manager_bal/bal.py [hunk 1]
```diff
@@ -343,9 +349,33 @@ def _copy_and_expand_trait_properties(entity_version_dict: dict, library: dict)
                     os.environ, **library["variables"]
                 )
 
+                subbed_val = trait_data[prop]
+
+                if subbed_val.startswith("file:/") or subbed_val.startswith("file:\\"):
```
Comment (feltech) [inline plugin/openassetio_manager_bal/bal.py:354]: [question] Are we saying values like `file:///some/path is the path to the entity`, i.e. where it's not _really_ a path, is just undefined behaviour?

File: README.md [hunk 2]
```diff
@@ -26,8 +26,12 @@ manager behavior.
   values (using the `$var` or `${var}` syntax, escape `$` using `$$`). A
   library can also define arbitrary variables of its own under the
   top-level `variables` key. In addition, BAL provides the built-in
-  `$bal_library_path` and `$bal_library_dir` variables, which can be
-  used to anchor to the current library location.
+  `$bal_library_path`, `$bal_library_dir` and `$bal_library_url`
```
Comment (feltech) [inline README.md:29]: [fix]
```suggestion
  `$bal_library_path`, `$bal_library_dir` and `$bal_library_dir_url`
```

**Review by feltech — CHANGES_REQUESTED** [review]
Looks good, really quite elegant, much more so than I anticipated, good work!  Just a few minor niggles.

File: tests/resources/library_business_logic_suite_var_expansion.json [hunk 3]
```diff
@@ -18,11 +18,15 @@
               "aBool": true,
               "bal_library_dir": "Library is in $bal_library_dir",
               "bal_library_path": "Library is ${bal_library_path}",
+              "bal_library_dir_url": "Library is in ${bal_library_dir_url}",
+              "relative_to_bal_library_dir_url": "${bal_library_dir_url}/../aboveFile.txt",
+              "relative_to_bal_library_dir_url_windows": "${bal_library_dir_url}\\..\\aboveFile.txt",
```
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply tests/resources/library_business_logic_suite_var_expansion.json:23]: I was assuming, (foolishly, as we didn't discuss it) windows style seperators would have been a requirement, that's why I split these tests, to test that explicitly.

If you're saying you'd rather not even allow them, I guess that simplifies things.

File: plugin/openassetio_manager_bal/bal.py (hunk 1 above)
Comment (elliotcmorris) [reply plugin/openassetio_manager_bal/bal.py]: I think we are, i wrote "url presenting" in the docs as a gesture to this. Not sure of an alternative that isn't too much effort/fiddly complexity for the value, given that using space as a delimeter is off the cards.

[the remark this answers is not in the document]
Comment (elliotcmorris) [reply plugin/openassetio_manager_bal/bal.py]: Ach, this is more me getting in a fritz over windows stuff than any real decision, I think probably just `file:` would be best, be as permissive as we can, gonna change to that.

File: plugin/openassetio_manager_bal/bal.py [hunk 4]
```diff
@@ -343,9 +349,33 @@ def _copy_and_expand_trait_properties(entity_version_dict: dict, library: dict)
...
+                    trait_data[prop] = normalize_file_url(subbed_val)
+
     return expanded_dict
 
 
+def normalize_file_url(maybe_file_url):
+    """
+    Any string in the library that begins with "file:///" will be
```
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply plugin/openassetio_manager_bal/bal.py]: As usual you assume too much of me, i didn't realise `/` style slashes _were_ good behaviour in URLS.

File: tests/bal_business_logic_suite.py [hunk 5]
```diff
@@ -670,6 +675,54 @@ def test_when_bal_library_dir_used_then_expanded_to_library_directory(self):
...
         self.assertEqual(data["bal_library_dir"], f"Library is in {expected_dir}")
 
+    def test_when_bal_library_url_used_then_expanded_to_library_directory(self):
+        expected_url = os.path.join(os.path.dirname(__file__), "resources")
+        expected_url = pathlib.Path(expected_url).resolve().as_uri()
+        data = self.__resolve_to_dict()
+        self.assertEqual(data["bal_library_dir_url"], f"Library is in {expected_url}")
+
+    def test_when_relative_bal_library_dir_used_then_library_path_not_normalized(self):
+        expected_url = os.path.join(os.path.dirname(__file__), "resources/../aboveFile.txt")
```
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply tests/bal_business_logic_suite.py:685]: Yeah, good idea.

File: README.md (hunk 2 above)
Comment (elliotcmorris) [reply README.md]: https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/pull/106/commits/b970aaa66b5f20579e193b82b792c14a302dd007

File: tests/bal_business_logic_suite.py [hunk 6]
```diff
@@ -670,6 +675,54 @@ def test_when_bal_library_dir_used_then_expanded_to_library_directory(self):
...
+
+    @unittest.skipIf(
+        sys.platform == "win32", reason="Constructing path expectations requires appropriate OS"
+    )
+    def test_bal_library_dir_path_from_url_compatible(self):
+        expected_path = os.path.join(os.path.dirname(__file__), "aboveFile.txt")
+        converter = FileUrlPathConverter()
+        data = self.__resolve_to_dict()
+        substituted_url = data["relative_to_bal_library_dir_url"]
+        path = converter.pathFromUrl(substituted_url, PathType.kPOSIX)
```
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply tests/bal_business_logic_suite.py]: Neat idea. I was under the impression we had to tests windows seperators in urls seperately/

https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/pull/106

File: tests/resources/library_business_logic_suite_var_expansion.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply tests/resources/library_business_logic_suite_var_expansion.json]: https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/pull/106

File: plugin/openassetio_manager_bal/bal.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply plugin/openassetio_manager_bal/bal.py]: https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/pull/106

File: plugin/openassetio_manager_bal/bal.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply plugin/openassetio_manager_bal/bal.py]: https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/pull/106

File: tests/resources/library_business_logic_suite_var_expansion.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply tests/resources/library_business_logic_suite_var_expansion.json]: https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/pull/106/commits/f046e5fff7e6d8f4e9e11536368bf674d044f3e2

File: plugin/openassetio_manager_bal/bal.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply plugin/openassetio_manager_bal/bal.py]: https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/pull/106/commits/f046e5fff7e6d8f4e9e11536368bf674d044f3e2

File: plugin/openassetio_manager_bal/bal.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply plugin/openassetio_manager_bal/bal.py]: https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/pull/106/commits/f046e5fff7e6d8f4e9e11536368bf674d044f3e2

File: plugin/openassetio_manager_bal/bal.py [hunk 7]
```diff
@@ -81,6 +85,7 @@ def load_library(path: str) -> dict:
             library.update(json.load(file))
         lib_path = os.path.abspath(path)
         lib_dir = os.path.dirname(lib_path)
+        lib_dir_url = pathlib.Path(pathlib.PurePosixPath(lib_dir)).as_uri()
```
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply plugin/openassetio_manager_bal/bal.py]: Leftover experimental cruft :upside_down_face: 

https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/pull/106/commits/7305408f9fdbab30013e96ff81106fdfb96a22cf

File: tests/bal_business_logic_suite.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply tests/bal_business_logic_suite.py:686]: https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/pull/106/commits/5cdbf707f4163f08bff4c148680791f224993aea

**Pull request merged**